File: base/android/statistics_recorder_android.cc
// Copyright 2016 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <string>

#include "base/android/jni_string.h"
#include "base/metrics/histogram_base.h"
#include "base/metrics/statistics_recorder.h"
#include "base/sys_info.h"
#include "jni/StatisticsRecorderAndroid_jni.h"

using base::android::JavaParamRef;
using base::android::ConvertUTF8ToJavaString;

namespace base {
namespace android {

static ScopedJavaLocalRef<jstring> ToJson(JNIEnv* env,
                                          const JavaParamRef<jclass>& clazz) {
  // Use lossy dump on low memory devices to reduce memory pressure, and
  // out-of-memory crashes in the browser process.
  base::JSONVerbosityLevel verbosity_level = base::JSONVerbosityLevel::kFull;
  if (base::SysInfo::IsLowEndDevice()) {
    verbosity_level = base::JSONVerbosityLevel::kSkipBuckets;
  }
  return ConvertUTF8ToJavaString(
      env, base::StatisticsRecorder::ToJSON(verbosity_level));
}

}  // namespace android
}  // namespace base

File: base/android/java/src/org/chromium/base/metrics/StatisticsRecorderAndroid.java
// Copyright 2016 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.base.metrics;

import org.chromium.base.annotations.JNINamespace;

/**
 * Java API which exposes the registered histograms on the native side as
 * JSON test.
 */
@JNINamespace("base::android")
public final class StatisticsRecorderAndroid {
    private StatisticsRecorderAndroid() {}

    /**
     * @param omitBuckets Set to true if the serialization of buckets should be omitted during the
     * serialization of the histograms. Should be set to true when it is critical to reduce the
     * memory pressure during the serialization.
     * @return All the registered histograms as JSON text.
     */
    public static String toJson(boolean omitBuckets) {
        return nativeToJson(omitBuckets);
    }

    private static native String nativeToJson(boolean omitBuckets);
}